tools/chplcheck/src/rules.py

```diff
@@ -323,3 +323,111 @@ def variables(node):
 
         for unused in indices.keys() - uses:
             yield indices[unused]
+
+    @driver.advanced_rule
+    def IncorrectIndentation(context, root):
+        """
+        Warn for inconsistent or missing indentation
+        """
+
+        # First, recurse and find warnings in children.
+        for child in root:
+            yield from IncorrectIndentation(context, child)
+
+        def contains_statements(node):
+            """
+            Returns true for allow-listed AST nodes that contain
+            just a list of statements.
+            """
+            return isinstance(node, (Record, Class, Module, Block))
+
+        def unwrap_intermediate_block(node):
+            """
+            Given a node, find the reference indentation that its children
+            should be compared against.
+
+            This method also rules out certain nodes that should
+            not be used for parent-child indentation comparison.
+            """
+            if not isinstance(node, Block):
+                return node
+
+            parent = node.parent()
+            if not parent:
+                return node
+
+            if isinstance(parent, (Function, Loop)):
+                return parent
+            elif isinstance(parent, Conditional) and parent.is_expression_level():
+                return None
+            return node
+
+        # If root is something else (e.g., function call), do not
+        # apply indentation rules; only apply them to things that contain
+        # a list of statements.
+        is_eligible_parent_for_indentation = contains_statements(root)
+        if not is_eligible_parent_for_indentation:
+            return
+
+        parent_for_indentation = unwrap_intermediate_block(root)
+        parent_depth = None
+        if parent_for_indentation is None:
+            # don't compare against any parent depth.
+            pass
+
+        # For implicit modules, proper code will technically be on the same
+        # line as the module's body. But we don't want to warn about that,
+        # since we don't want to ask all code to be indented one level deeper.
+        elif not (isinstance(parent_for_indentation, Module) and parent_for_indentation.kind() == "implicit"):
+            parent_depth = parent_for_indentation.location().start()[1]
+
+        prev = None
+        prev_depth = None
+        prev_line = None
+
+        # We only care about misaligned statements, so we don't want to do stuff
+        # like warn for inherit-exprs or pragmas on a record.
+        iterable = root
+        if isinstance(root, AggregateDecl):
+            iterable = root.decls_or_comments()
+
+        for child in iterable:
+            if isinstance(child, Comment): continue
+
+            # NamedDecl nodes currently use the name as the location, which
+            # does not indicate their actual indentation.
+            if isinstance(child, (NamedDecl, TupleDecl, ForwardingDecl)): continue
```

Comment:
Comment is incorrect, some NamedDecl use the full body. This should be `VarLikeDecl`.